ApiData.ts

```diff
@@ -0,0 +1,264 @@
+					this.displayRecords();
+					this.updatePageInfo();
+					// empty search input after searching 
+					$('#fromInput').val('');
+				})
+				.catch(() => {
+					throw ('Failed to search value');
+				});
+		} else {
+			alert(`Error while searching , please enter values in the range (0-${maxRange})`);
+			return Promise.resolve();
+		}
+	}
+
+	/** use Ajax for data fetching */
+	private async fetchData(url: string): Promise<number | string> {
+		$('#overlay').show();
+		const response = await $.ajax({
+			url,
+			method: 'GET',
+		});
+		$('#overlay').hide();
+		return response;
+	}
+
+	/** Change grid height according to screen size */
+	private adjustGridHeight(): void {
+		const gridElement = document.getElementById('grid');
+		const pageCntrl = $('.grid-controls').innerHeight();
+		const screenHeight = $(window).innerHeight();
+		if (gridElement && pageCntrl !== undefined && screenHeight !== undefined) {
+			this.maxGridHeight = screenHeight - pageCntrl;
+			gridElement.style.height = `${this.maxGridHeight}px`;
+		}
+	}
+
+	/** Update the page information and records display based on the current state of the grid. */
+	private updatePageInfo(): void {
+		const totalPages = Math.ceil(this.totalItems / this.pageSize);
+		const pageInfo = `Page ${this.currentPage} of ${totalPages}`;
+		const from = this.firstVal;
+		let to = Math.min(from + this.pageSize, this.maxRange);
+		$('#pageInfo').text(`${pageInfo}`);
+		$('.records').text(`Showing records ${from} to ${to}`);
+	}
+
+	private setupControls(): void {
+		$('#prevBtn').on('click', () => this.handlePageChange(-1));
+		$('#nextBtn').on('click', () => this.handlePageChange(1));
+		$(window).on('resize', debounce(() => { this.handleResize(); }, 100));
+	}
+
+	/** Handles page navigation by updating the firstVal, lastVal, current page, and enabling/disabling previous and next buttons as needed. */
+	private handlePageChange(delta: number): void {
+		let prevBtn = $('#prevBtn');
+		let nextBtn = $('#nextBtn');
+
+		// Check if delta(change in page number) is positive and disable the next page if firstval + pageSize exceeds the MaxRange.
+		if (delta > 0 && this.firstVal + delta * this.pageSize > this.maxRange) {
+			this.firstVal = this.lastVal - this.pageSize;
+			prevBtn.attr("disabled", null);
+			nextBtn.attr("disabled", "disabled");
+		} else if (delta < 0 && this.firstVal + delta * this.pageSize < 0) {
+			// If delta is negative then reset firstVal to 0 and disable prev button 
+			this.firstVal = 0;
+			prevBtn.attr("disabled", "disabled");
+			nextBtn.attr("disabled", null);
+		} else {
+			this.firstVal = Math.max(0, Math.min(this.firstVal + delta * this.pageSize, this.maxRange));
+			prevBtn.attr("disabled", null);
+			nextBtn.attr("disabled", null);
+		}
+
+		this.lastVal = this.firstVal + delta * this.pageSize;
+		this.currentPage = Math.floor(this.firstVal / this.pageSize) + 1;
+
+		this.fetchAndDisplayRecords()
+			.then(() => {
+				this.updatePageInfo();
```

Comment:
Please look at the number of times you call `this.updatePageInfo()` and see if it really is necessary? I would think fetchAndDisplayRecords() already called it for you and `this.displayRecords()` as well and now `this.handlePageChange()` as well.